### DISCUSSION TITLE
Reusable array shapes

### DISCUSSION THREAD

**Discussion by juacala** [vimeo/psalm#7869] [Ideas]
One thing I run into is an array shape that is reused in a lot of places. These are typically associative arrays from a database table, and is passed around to various functions. 

What is the best way to annotate such an array shape in one place and reuse it, so you don’t need to rewrite the same shape in lots of function docblocks?

One thought I had would be to make a plug-in that respects syntax like
@param array{\path\to\some\Class}
and define the properties in a class definition. 

The nice thing about that would be, once psalm validates all the usage, you could migrate it to use the class instead of the array. 

I’m interested to know how others deal with this.

**Comment by weirdan** [comment]
`@psalm-type` + `@psalm-import-type` (https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-import-type)
